// pg-protocol/src/testing/buffer-list.ts
export default class BufferList {
  constructor(public buffers: Buffer[] = []) {}

  public add(buffer: Buffer, front = false) {
    this.buffers[front ? 'unshift' : 'push'](buffer)
    return this
  }

  public addInt16(val: number, front = false) {
    return this.add(Buffer.from([val >>> 8, val >>> 0]), front)
  }

  public addInt32(val: number, first = false) {
    return this.add(
      Buffer.from([(val >>> 24) & 0xff, (val >>> 16) & 0xff, (val >>> 8) & 0xff, (val >>> 0) & 0xff]),
      first
    )
  }

  public addCString(val: string, front = false) {
    const len = Buffer.byteLength(val)
    const buffer = Buffer.alloc(len + 1)
    buffer.write(val, 0, len, 'utf8')
    buffer[len] = 0
    return this.add(buffer, front)
  }

  public addString(val: string, front = false) {
    const len = Buffer.byteLength(val)
    const buffer = Buffer.alloc(len)
    buffer.write(val, 0, len, 'utf8')
    return this.add(buffer, front)
  }

  public addChar(char: string, first = false) {
    return this.add(Buffer.from(char, 'utf8'), first)
  }

  public addByte(byte: number) {
    return this.add(Buffer.from([byte]))
  }

  public join(appendLength = false, char?: string): Buffer {
    const length = this.getByteLength()
    if (appendLength) {
      this.addInt32(length + 4, true)
      return this.join(false, char)
    }
    if (char) {
      this.addChar(char, true)
      length++
    }
    const result = Buffer.alloc(length)
    let index = 0
    for (const buffer of this.buffers) {
      buffer.copy(result, index, 0)
      index += buffer.length
    }
    return result
  }

  public static concat(...buffers: Buffer[]): Buffer {
    const total = new BufferList()
    for (const buffer of buffers) {
      total.add(buffer)
    }
    return total.join()
  }

  private getByteLength(initial = 0): number {
    return this.buffers.reduce((previous, current) => previous + current.length, initial)
  }
