// pg-protocol/dist/buffer-reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BufferReader = void 0;

class BufferReader {
    constructor(buffer = Buffer.allocUnsafe(0), offset = 0) {
        this.buffer = buffer;
        this.offset = offset;
        this.encoding = 'utf-8';
    }

    setBuffer(offset, buffer) {
        this.offset = offset;
        this.buffer = buffer;
    }

    readInt16BE() {
        const result = this.buffer.readInt16BE(this.offset);
        this.offset += 2;
        return result;
    }

    readByte() {
        const result = this.buffer[this.offset];
        this.offset++;
        return result;
    }

    readInt32BE() {
        const result = this.buffer.readInt32BE(this.offset);
        this.offset += 4;
        return result;
    }

    readString(length) {
        const result = this.buffer.toString(this.encoding, this.offset, this.offset + length);
        this.offset += length;
        return result;
    }

    readCstring() {
        let end = this.offset;
        while (this.buffer[end++] !== 0) { }
        const result = this.buffer.toString(this.encoding, this.offset, end - 1);
        this.offset = end;
        return result;
    }

    readBytes(length) {
        const result = this.buffer.slice(this.offset, this.offset + length);
        this.offset += length;
        return result;
    }
}
exports.BufferReader = BufferReader;
